### PR TITLE
Cubical cover

### PR DESCRIPTION
**Pull request by lewtun**
#### What does this implement/fix? Explain your changes.
* Implements a simple cover of an n-dimensional box by taking products of intervals of the covers of [min_n, max_n] for each of the n lenses.
* Refactors `My1DCover` to have class name and constructor variables that better reveal intent

### CODE REVIEW

File: giotto/mapper/cover.py [hunk 1]
```diff
@@ -1,11 +1,12 @@
 import numpy as np
 from sklearn.base import BaseEstimator, TransformerMixin
+from itertools import product
 
 
-class My1DCover(BaseEstimator, TransformerMixin):
-    def __init__(self, n_intervals=10, overlap=0.1):
+class OneDimensionalCover(BaseEstimator, TransformerMixin):
+    def __init__(self, n_intervals=10, overlap_pct=0.1):
```
Comment (ulupo) [inline giotto/mapper/cover.py]: Not really a percentage, more of a fractional quantity ("frac_overlap", "overlap_ratio", ...)

File: giotto/mapper/cover.py [hunk 2]
```diff
@@ -14,13 +15,50 @@ def transform(self, X, y=None):
...
+            intervals_count = cover.shape[1]
+            intervals.append(
+                [cover[:, interval_index]
+                 for interval_index in range(intervals_count)]
+            )
+
+        # calculate masks for pullback cover
+        Xt = np.array([np.logical_and.reduce(t)
+                       for t in product(*intervals)]).T
+        return Xt
```
Comment (ulupo) [inline giotto/mapper/cover.py]: Could clean up any purely False arrays before returning Xt

File: giotto/mapper/cover.py [hunk 3]
```diff
@@ -14,13 +15,50 @@ def transform(self, X, y=None):
...
+
+        for lens_index in range(lens_count):
+            lens = X[:, lens_index, None]
+            cover = OneDimensionalCover(
+                n_intervals=self.n_intervals, overlap_pct=self.overlap_pct)
+            covers.append(cover.fit_transform(lens))
+
+        intervals = []
+
+        for cover in covers:
```
Comment (ulupo) [inline giotto/mapper/cover.py]: Might a list comprehension be faster without too much loss in readability?

File: giotto/mapper/cover.py [hunk 4]
```diff
@@ -14,13 +15,50 @@ def transform(self, X, y=None):
...
         Xt = (X > left_endpts) & (X < right_endpts)
         return Xt
+
+
+class CubicalCover(BaseEstimator, TransformerMixin):
+    """Calculates the cover of an n-dimensional hypercube by taking products
+    of intervals of the covers of [min_n, max_n] for each of the n lenses.
+    """
+
+    def __init__(self, n_intervals, overlap_pct):
```
Comment (ulupo) [inline giotto/mapper/cover.py]: Ditto

**Comment by lewtun** [comment]
@ulupo I implemented your suggestions. Please review and merge if acceptable
